source/components/AblyTask/AblyTask.brs

```diff
@@ -9,53 +12,78 @@ sub init()
     ATTACHED: 11,
     MESSAGE: 15
   }
-  m.headers = {
-    "Accept": "application/json"
-    "Content-Type": "application/json"
-  }
 
+  ' Highest supported log level. If set higher it will be lowered to match.
   m.MAXIMUM_LOG_LEVEL = 5
 end sub
 
+' Any code run from this functions will be run in the async task thread
 sub runTask()
+  ' Get all the public values in one pass to limit the amount of rendezvous
   topValues = m.top.getFields()
+
+  ' Process constructor values
   m.ENDPOINT = topValues.endpoint
-  m.CHANNEL = topValues.channel
-  m.KEY = getConnectionKey()
+  m.CHANNELS = topValues.channels
   m.logLevel = m.top.logLevel
   if m.logLevel > m.MAXIMUM_LOG_LEVEL then m.logLevel = m.MAXIMUM_LOG_LEVEL
 
-  logInfo(m.KEY)
+  ' Get the Authentication key
+  m.authenticationKey = getAuthenticationKey()
+
+  ' Make sure we got something back to use as am authentication key
+  logInfo("Authentication Key:", m.authenticationKey)
   m.connectionKey = ""
-  if NOT isNonEmptyString(m.KEY) then return
+  if NOT isNonEmptyString(m.authenticationKey) then return
 
+  ' Establish the initial connection
   if connect() then
-    if attach() then
-      while stream()
-        sleep(20)
-      end while
-    end if
+    for each channel in m.CHANNELS
+      ' Send an attach requests for each channel
+      logInfo("Attaching:", channel, attach(channel))
+    end for
+
+    ' Start watching for events
+    while stream()
```

Comment:
I would either rename the function `stream()` to `receiveEvent()` or something like that (more descriptive) or leave the name as is and just move the `while` loop into the function. I can't see any benefit of pulling that loop up here.

Also, I'm not sure there's likely any evidence you actually need the `sleep(20)` between each call to poll the server. Generally the polls themselves will provide plenty of idle time as they await a response from the service. I'm not sure you can have any evidence this is required at this stage so this smells like a premature optimisation.

Reply:
Most times we have one main event loop in the main task function. I agree with renaming it for sure.

As for the overheat. I added this as this is something I have run into with a few tasks I've worked on. This will likely get reworked at some point to be non-blocking and once that happens the sleep would be removed as a side effect.

Reply:
Saw your change https://github.com/ably/ably-roku/pull/5/commits/ecadc9e9b8ace484857e37335840f0685600e401, but would still argue there's little need for that empty while loop at this high level. I would just move the loop in to the function and perhaps revert to naming it `stream()` per my suggestion at the start of this conversation.

Reply:
fd5bc0a